### PR TITLE
Add Fast Image Processor for ImageGPT

### PR DESCRIPTION
**Pull request by agamjots05**
# What does this PR do?

Related #36978

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @yonigozlan

**Comment by yonigozlan** [comment]
Hey @agamjots05 and @ayaayethan ! Thanks for opening this PR, don't hesitate to ping me when it's ready to review or if you need help!

**Comment by agamjots05** [comment]
> Hey @agamjots05 and @ayaayethan ! Thanks for opening this PR, don't hesitate to ping me when it's ready to review or if you need help!

Thank you for letting us know! We'll let you know if we get stuck on anything 👍

**Comment by agamjots05** [comment]
@yonigozlan Quick question
My team and I have made the fast processor for ImageGPT and have come across an interesting issue in regards to the testing file for ImageGPT.

**To get the final two slow/fast parity tests to pass, we needed to make two very small changes in the slow ImageGPTImageProcessor:**
1. Always include pixel_values in the returned BatchFeature even when do_color_quantize=True (alongside input_ids).
This avoids KeyError/AttributeError in the shared test mixins that access encoding.pixel_values. It’s additive-only.
2. Override to_dict() to convert clusters from np.ndarray to a plain Python list.
This makes the config JSON-serializable, which allows us to pass a test we were failing 

**Would these two minimal adjustments to the slow processor be acceptable?**
If not should we, for the time being, ignore these failing test cases occurring due to the base processor?

**Comment by yonigozlan** [comment]
Hey @agamjots05 !
These seems like reasonable changes, strange that it didn't cause any issues before, thanks for fixing!

**Comment by agamjots05** [comment]
> Hey @agamjots05 ! These seems like reasonable changes, strange that it didn't cause any issues before, thanks for fixing!

Sounds good. If that's the case, then this PR should be ready for review 👍

File: src/transformers/models/imagegpt/image_processing_imagegpt.py [hunk 1]
```diff
@@ -291,14 +297,21 @@ def preprocess(
...
             images = list(images)
+            data = {"input_ids": images, "pixel_values": pixel_values}
         else:
             images = [
                 to_channel_dimension_format(image, data_format, input_channel_dim=input_data_format)
                 for image in images
             ]
-
-        data = {"input_ids": images}
+            data = {"pixel_values": images}
```
Comment (yonigozlan) [inline src/transformers/models/imagegpt/image_processing_imagegpt.py]: Actually I did not understand your previous message. If the original image processor output contained only an "input_ids" key in all cases, we should keep it that way, and have the same for the fast image processor

File: src/transformers/models/imagegpt/image_processing_imagegpt_fast.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+        image is used. Can be one of:
+            - `"channels_first"` or `ChannelDimension.FIRST`: image in (num_channels, height, width) format.
+            - `"channels_last"` or `ChannelDimension.LAST`: image in (height, width, num_channels) format.
+            - `"none"` or `ChannelDimension.NONE`: image in (height, width) format.
+    input_data_format (`ChannelDimension` or `str`, *optional*):
+        The channel dimension format for the input image. If unset, the channel dimension format is inferred
+        from the input image. Can be one of:
+            - `"channels_first"` or `ChannelDimension.FIRST`: image in (num_channels, height, width) format.
+            - `"channels_last"` or `ChannelDimension.LAST`: image in (height, width, num_channels) format.
+            - `"none"` or `ChannelDimension.NONE`: image in (height, width) format.
```
Comment (yonigozlan) [inline src/transformers/models/imagegpt/image_processing_imagegpt_fast.py]: No need for all these here, return_tensors is already in the base fast image processors, and no need for data formats args, it's always channel first for fast image processors.

File: src/transformers/models/imagegpt/image_processing_imagegpt_fast.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+
+@auto_docstring
+class ImageGPTImageProcessorFast(BaseImageProcessorFast):
+    """
+    Constructs a fast ImageGPT image processor.
+
+    This processor can be used to resize images to a smaller resolution (such as 32x32 or 64x64),
+    normalize them and finally color quantize them to obtain sequences of "pixel values" (color clusters).
+    """
```
Comment (yonigozlan) [inline src/transformers/models/imagegpt/image_processing_imagegpt_fast.py]:
```suggestion
```
Docstring is handled by @auto_docstring

File: src/transformers/models/imagegpt/image_processing_imagegpt_fast.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+
+    model_input_names = ["input_ids"]
+
+    # Defaults largely aligned with the slow processor, except normalization which we do manually to [-1, 1]
+    resample = PILImageResampling.BILINEAR
+    size = {"height": 256, "width": 256}
+    do_resize = True
+    # We do NOT want touse the base normalization/rescale as ImageGPT expects (x/127.5 - 1)
+    do_rescale = False
+    do_normalize = False
```
Comment (yonigozlan) [inline src/transformers/models/imagegpt/image_processing_imagegpt_fast.py]: I'm sure you can have the same results using typical normalization and rescaling. using image_mean =[0.5, 0.5, 0.5], image_std=[0.5, 0.5, 0.5] and keeping the default rescale factor (1/255) should be equivalent

File: src/transformers/models/imagegpt/image_processing_imagegpt_fast.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+
+    do_color_quantize = True
+    clusters = None
+
+    def __init__(
+        self,
+        clusters: Optional[Union[list, np.ndarray]] = None,
```
Comment (yonigozlan) [inline src/transformers/models/imagegpt/image_processing_imagegpt_fast.py]: Add that we keep this as an arg for backward compatibility reasons

File: src/transformers/models/imagegpt/image_processing_imagegpt_fast.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+        **kwargs: Unpack[ImageGPTFastImageProcessorKwargs],
+    ):
+        super().__init__(**kwargs)
+        # Store clusters as numpy for JSON serializability. Convert to torch in _preprocess when needed.
+        if clusters is not None:
+            self.clusters = np.asarray(clusters, dtype=np.float32)
+        else:
+            self.clusters = None
```
Comment (yonigozlan) [inline src/transformers/models/imagegpt/image_processing_imagegpt_fast.py]: why not define as torch tensor directly and convert from torch to list in to_dict?
```suggestion
        self.clusters = torch.tensor(clusters, dtype=torch.float32) if clusters is not None else None
```

File: src/transformers/models/imagegpt/image_processing_imagegpt_fast.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+        # Default: follow ImageGPT behavior (normalize by default). We stash here and force base to skip.
+        self._do_normalize_imagegpt = kwargs.get("do_normalize", True)
```
Comment (yonigozlan) [inline src/transformers/models/imagegpt/image_processing_imagegpt_fast.py]: as said above, let's use typical normalization rescaling

File: src/transformers/models/imagegpt/image_processing_imagegpt_fast.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+
+    def _further_process_kwargs(self, **kwargs):
+        # Let the base process size/crop and other standard kwargs first
+        kwargs = super()._further_process_kwargs(**kwargs)
+        if "do_normalize" in kwargs and kwargs["do_normalize"] is not None:
+            self._do_normalize_imagegpt = kwargs["do_normalize"]
+        # Force base pipeline to skip its rescale/normalize validation and logic
+        kwargs["do_rescale"] = False
+        kwargs["do_normalize"] = False
+        return kwargs
```
Comment (yonigozlan) [inline src/transformers/models/imagegpt/image_processing_imagegpt_fast.py]:
```suggestion
```
No need for that then

File: src/transformers/models/imagegpt/image_processing_imagegpt_fast.py [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+            clusters (`np.ndarray`, `list[list[float]]`, or `torch.Tensor`, *optional*, defaults to `self.clusters`):
+                Clusters used to quantize the image of shape `(n_clusters, 3)`. Only has an effect if
+                `do_color_quantize` is set to `True`.
+            return_tensors (`str` or `TensorType`, *optional*):
+                The type of tensors to return. Can be one of:
+                    - Unset: Return a list of `torch.Tensor`.
+                    - `TensorType.PYTORCH` or `'pt'`: Return a batch of type `torch.Tensor`.
+                    - `TensorType.NUMPY` or `'np'`: Return a batch of type `np.ndarray`.
+                    - `TensorType.JAX` or `'jax'`: Return a batch of type `jax.numpy.ndarray`.
+        """
```
Comment (yonigozlan) [inline src/transformers/models/imagegpt/image_processing_imagegpt_fast.py]: docstrings are already above, let's not clutter the function with them
```suggestion
```

File: src/transformers/models/imagegpt/image_processing_imagegpt_fast.py [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+        # Run standard fast pipeline (resize, crop, batching) without rescale/normalize
+        base_batch = super()._preprocess(images, return_tensors=return_tensors, **kwargs)
```
Comment (yonigozlan) [inline src/transformers/models/imagegpt/image_processing_imagegpt_fast.py]: Let's unroll the standard fast pipeline here, so it's simpler for users checking the code to see what processing is really being done

File: src/transformers/models/imagegpt/image_processing_imagegpt_fast.py [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+        kwargs["do_normalize"] = False
+        return kwargs
+
+    def _preprocess(
+        self,
+        images,
+        do_color_quantize: Optional[bool] = None,
+        clusters: Optional[Union[list, np.ndarray, torch.Tensor]] = None,
+        return_tensors: Optional[str] = None,
+        **kwargs: Unpack[ImageGPTFastImageProcessorKwargs],
```
Comment (yonigozlan) [inline src/transformers/models/imagegpt/image_processing_imagegpt_fast.py]: let's unroll the kwargs in `_preprocess`

File: src/transformers/models/imagegpt/image_processing_imagegpt_fast.py [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+        # Apply ImageGPT normalization when requested: [-1, 1]
+        do_normalize = getattr(self, "_do_normalize_imagegpt", True)
+        if isinstance(pixel_values, torch.Tensor):
+            normalized = pixel_values.to(dtype=torch.float32)
+            if do_normalize:
+                normalized = F.normalize(normalized, mean=[0.0], std=[127.5]) - 1.0
+        else:
+            normalized = [img.to(dtype=torch.float32) for img in pixel_values]
+            if do_normalize:
+                normalized = [F.normalize(img, mean=[0.0], std=[127.5]) - 1.0 for img in normalized]
```
Comment (yonigozlan) [inline src/transformers/models/imagegpt/image_processing_imagegpt_fast.py]: let's use typical rescale/normalization

File: src/transformers/models/imagegpt/image_processing_imagegpt_fast.py [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+            from ...image_processing_utils import BatchFeature
+
+            return BatchFeature(
+                data={"input_ids": input_ids, "pixel_values": pixel_values_out},
+                tensor_type=return_tensors,
+            )
+
+        # Otherwise, return pixel values (normalized or not depending on flag)
+        base_batch["pixel_values"] = normalized
+        return base_batch
```
Comment (yonigozlan) [inline src/transformers/models/imagegpt/image_processing_imagegpt_fast.py]: This needs to be redone fully, let's do quantization by batch (by grouping by shape like for other operations in the base fast image proc. if needed), no function definition inside other functions, import should be at the top.

File: src/transformers/models/imagegpt/image_processing_imagegpt_fast.py [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+
+    def to_dict(self):
+        # Convert numpy arrays to lists for JSON serialization
+        output = super().to_dict()
+        if output.get("clusters") is not None and isinstance(output["clusters"], np.ndarray):
+            output["clusters"] = output["clusters"].tolist()
```
Comment (yonigozlan) [inline src/transformers/models/imagegpt/image_processing_imagegpt_fast.py]: you can do the same with torch tensors

File: src/transformers/models/imagegpt/image_processing_imagegpt_fast.py [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+
+        # Need to set these valus to match with slow processor during testing
+        output["rescale_factor"] = None
+        output["do_rescale"] = None
+        output["do_color_quantize"] = bool(getattr(self, "do_color_quantize", True))
+        output.pop("_do_normalize_imagegpt", None)
```
Comment (yonigozlan) [inline src/transformers/models/imagegpt/image_processing_imagegpt_fast.py]: this should not be necessary

File: tests/models/imagegpt/test_image_processing_imagegpt.py [hunk 16]
```diff
@@ -181,69 +189,72 @@ def test_init_without_params(self):
...
+            # Test not batched input
+            encoded_images = image_processing(image_inputs[0], return_tensors="pt").input_ids
+            self.assertEqual(tuple(encoded_images.shape), (1, *expected_output_image_shape))
+
+            # Test batched
+            encoded_images = image_processing(image_inputs, return_tensors="pt").input_ids
+            self.assertEqual(
+                tuple(encoded_images.shape),
+                (self.image_processor_tester.batch_size, *expected_output_image_shape),
+            )
```
Comment (yonigozlan) [inline tests/models/imagegpt/test_image_processing_imagegpt.py:268]: You'll probably need to override the two test_slow_fast_equivalence tests since the outputs will have input_ids as its only attribute

**Review by yonigozlan — COMMENTED** [review]
Hey @agamjots05 and @ayaayethan , thanks for contributing! There's a few changes to do, and I'll be off for two weeks, but happy to do another review when I'm back!

**Comment by agamjots05** [comment]
@yonigozlan I believe I have implemented all the changes you have requested. Please take a look when you can 👍

**Review by yonigozlan — APPROVED** [review]
Thanks for working on this @agamjots05 and @ayaayethan !
I just pushed some small simplifications to the color quantize logic, and I'll merge once the CI passes!
